shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.mm

```diff
@@ -458,58 +462,51 @@ @interface FlutterClippingMaskViewPool ()
 // The maximum number of `FlutterClippingMaskView` the pool can contain.
 // This prevents the pool to grow infinately and limits the maximum memory a pool can use.
 @property(assign, nonatomic) NSUInteger capacity;
-@property(retain, nonatomic) NSMutableArray<FlutterClippingMaskView*>* pool;
-// The index points to the first available FlutterClippingMaskView in the `pool`.
-@property(assign, nonatomic) NSUInteger availableIndex;
+
+// The pool contains the views that are available to use.
+// The number of items in the pool must not excceds `capacity`.
+@property(retain, nonatomic) NSMutableSet<FlutterClippingMaskView*>* pool;
 
 @end
 
 @implementation FlutterClippingMaskViewPool : NSObject
 
 - (instancetype)initWithCapacity:(NSInteger)capacity {
   if (self = [super init]) {
-    _pool = [[NSMutableArray alloc] initWithCapacity:capacity];
+    // Most of cases, there are only one PlatformView in the scene.
+    // Thus init with the capacity of 1.
+    _pool = [[NSMutableSet alloc] initWithCapacity:1];
     _capacity = capacity;
-    _availableIndex = 0;
   }
   return self;
 }
 
 - (FlutterClippingMaskView*)getMaskViewWithFrame:(CGRect)frame {
-  FML_DCHECK(self.availableIndex <= self.capacity);
-  FlutterClippingMaskView* maskView;
-  if (self.availableIndex == self.capacity) {
-    // The pool is full, alloc a new one.
-    maskView =
+  FML_DCHECK(self.pool.count <= self.capacity);
+  if (self.pool.count == 0) {
+    // The pool is empty, alloc a new one.
+    return
         [[[FlutterClippingMaskView alloc] initWithFrame:frame
                                             screenScale:[UIScreen mainScreen].scale] autorelease];
-    return maskView;
   }
-
-  if (self.availableIndex >= self.pool.count) {
-    // The pool doesn't have enough maskViews, alloc a new one and add to the pool.
-    maskView =
-        [[[FlutterClippingMaskView alloc] initWithFrame:frame
-                                            screenScale:[UIScreen mainScreen].scale] autorelease];
-    [self.pool addObject:maskView];
-    FML_DCHECK(self.pool.count <= self.capacity);
-  } else {
-    // Reuse a maskView from the pool.
-    maskView = [self.pool objectAtIndex:self.availableIndex];
-    maskView.frame = frame;
-    [maskView reset];
-  }
-  self.availableIndex++;
+  FlutterClippingMaskView* maskView = [[[self.pool anyObject] retain] autorelease];
+  maskView.frame = frame;
+  [maskView reset];
+  [self.pool removeObject:maskView];
   return maskView;
 }
 
-- (void)recycleMaskViews {
-  self.availableIndex = 0;
+- (void)insertViewToPoolIfNeeded:(FlutterClippingMaskView*)maskView {
+  FML_DCHECK(![self.pool containsObject:maskView]);
+  FML_DCHECK(self.pool.count <= self.capacity);
```

Comment:
How is this safe? This method can be called from outside this class (since it's in an internal header), but I don't see a way to get the current pool size. How would a client avoid triggering this DCHECK?

Reply:
Oh, I missed the `=` part of the comparison. Sorry about that!